README.md
BS-Seeker2
=========

**BS-Seeker2 is a seamless and versatile pipeline for accurately and fast mapping the bisulfite-treated reads.**

**[CGmapTools](https://cgmaptools.github.io) is suggested for downstream analysis of BS-Seeker2.**  

---

  [Homepage](http://pellegrini.mcdb.ucla.edu/BS_Seeker2/) 
| [Mirror](http://guoweilong.github.io/BS_Seeker2/index.html) 
| [Published Paper](http://www.biomedcentral.com/1471-2164/14/774) 
| [Source code](https://github.com/BSSeeker/BSseeker2) 
| [Galaxy Toolshed](http://toolshed.g2.bx.psu.edu/repository?repository_id=e435334e4e9e19c1) 
| [UCLA Galaxy](http://galaxy.hoffman2.idre.ucla.edu)
| [CGmapTools](https://cgmaptools.github.io)


---

# Contents


1. [Remarkable new features](#1-remarkable-new-features)
2. [Supports](#2-supports)
3. [System requirements](#3-system-requirements)
4. [Module descriptions](#4-module-descriptions)

	* 4.1 [FilterReads.py](#41-filterReadspy)
	* 4.2 [bs_seeker2-build.py](#42-bs_seeker2-buildpy)
	* 4.3 [bs_seeker2-align.py](#43-bs_seeker2-alignpy)
	* 4.4 [bs_seeker2-call_methylation.py](#44-bs_seeker2-call_methylationpy)
	
5. [Contact Information](#5-contact-information)
6. [Questions and Answers](#6-questions-and-answers)

	* (1) [Performance](#1-performance)
	* (2) [Input/Output formats](#2--inputoutput-formats)
	* (3) ["Pysam" package related problem](#3-pysam-package-related-problem)
	* (4) [Configuration of BS-Seeker2](#4-configuration-of-bs-seeker2)
	* (5) [Unique alignment](#5-unique-alignment)
	* (6) [Paired-end sequencing alignment](#6-paired-end-sequencing-alignment)
	* (7) [Adapter related issue](#7-adapter-related-issue)
	* (8) [Others](#8-others)

# 1 Remarkable new features


* Reduced index for RRBS, accelerating the mapping speed and increasing mappability
* Allowing local/gapped alignment with Bowtie2, increased the mappability
* Option for removing reads suffering from bisulfite conversion failure

# 2 Supports

* Supported library types
	- Whole Genome-wide Bisulfite Sequencing (WGBS)
	- Reduced Representative Bisulfite Sequencing (RRBS)

* Supported formats for input file
	- [fasta](http://en.wikipedia.org/wiki/FASTA_format)
	- [fastq](http://en.wikipedia.org/wiki/FASTQ_format)
	- [qseq](http://jumpgate.caltech.edu/wiki/QSeq)
	- pure sequence (one-line one-sequence)

* Supported alignment tools
	- [bowtie](http://bowtie-bio.sourceforge.net/index.shtml) : Single-seed, fast, (default)
	- [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) : Multiple-seed, gapped-alignment
		- [local alignment](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#local-alignment-example) (default for bowtie2)
		- [end-to-end alignment](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#end-to-end-alignment-example)
	- [soap](http://soap.genomics.org.cn/)

* Supported formats for mapping results
	- [BAM](http://genome.ucsc.edu/FAQ/FAQformat.html#format5.1)
	- [SAM](http://samtools.sourceforge.net/)
	- [BS-seeker](http://pellegrini.mcdb.ucla.edu/BS_Seeker/USAGE.html)

# 3 System requirements

* Linux/Unix or Mac OS platform

* One of the following short read aligners

  [bowtie](http://bowtie-bio.sourceforge.net/), [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/), [soap](http://soap.genomics.org.cn/)

* [Python](http://www.python.org/download/) (Version 2.6 +)

  It is normally pre-installed in Linux. Type " python -V" to see the installed version.

* [pysam](http://code.google.com/p/pysam/) package (Version 0.6.x).

  Read "Questions & Answers" if you have problem when installing this package.

# 4 Module descriptions


## 4.1 FilterReads.py

Optional and independent module. **Not necessary.**
Some reads would be extremely amplified during the PCR. This script helps you get unique reads before doing the mapping.
You can decide whether or not to filter reads before doing the mapping.

* Usage :

```bash

	$ python FilterReads.py
    Usage: FilterReads.py -i <input> -o <output> [-k]
    Author : Guo, Weilong; guoweilong@gmail.com
    Start from: 2012-11-10; Last Update: 2017-12-08
    Description: Unique reads for qseq/fastq/fasta/sequence.
       Low quality reads in qseq file can be filtered.
    Warning: This function is reserved for WGBS, but not for RRBS.
    For WGBS, user can also try 'samtools rmdup' to get unique reads using BAM files.
    For RRBS, it is suggested not to get unique reads, as the starting ends of reads
    are more likely to be same for the reads from one C-CCG~~~C-CGG fragment.

    Options:
    -h, --help  show this help message and exit
    -i FILE     Name of the input qseq/fastq/fasta/sequence file
    -o FILE     Name of the output file
    -k          Would not filter low quality reads if specified, only applied
              for qseq format

```


* Tip :

> This step is not suggested for RRBS library, as reads from RRBS library would more likely from the same location.


## 4.2 bs_seeker2-build.py

Module to build the index for BS-Seeker2.


* Usage :


```bash

    $ python bs_seeker2-build.py -h
    
    Usage: bs_seeker2-build.py [options]

    Options:
      -h, --help            show this help message and exit
      -f FILE, --file=FILE  Input your reference genome file (fasta)
      --aligner=ALIGNER     Aligner program to perform the analysis: bowtie,
                            bowtie2, soap, rmap [Default: bowtie]
      -p PATH, --path=PATH  Path to the aligner program. Detected:
                            bowtie: /Install/bowtie-1.1.2/
                            bowtie2: /Install/bowtie2-master/
                            rmap: None
                            soap: None
      -d DBPATH, --db=DBPATH
                            Path to the reference genome library (generated in
                            preprocessing genome) [Default: /Install/BSseeker2/bs_utils/reference_genomes]
      -v, --version         show version of BS-Seeker2

      Reduced Representation Bisulfite Sequencing Options:
        Use this options with conjuction of -r [--rrbs]

        -r, --rrbs          Build index specially for Reduced Representation
                            Bisulfite Sequencing experiments. Genome other than
                            certain fragments will be masked. [Default: False]
        -l LOW_BOUND, --low=LOW_BOUND
                            lower bound of fragment length (excluding recognition
                            sequence such as C-CGG) [Default: 20]
        -u UP_BOUND, --up=UP_BOUND
                            upper bound of fragment length (excluding recognition
                            sequence such as C-CGG ends) [Default: 500]
        -c CUT_FORMAT, --cut-site=CUT_FORMAT
                            Cut sites of restriction enzyme. Ex: MspI(C-CGG),
                            Mael:(C-TAG), double-enzyme MspI&Mael:(C-CGG,C-TAG).
                            [Default: C-CGG]
```

* Example


```bash

    # Build genome index for WGBS using bowtie, path of bowtie should be included in $PATH
    python bs_seeker2-build.py -f genome.fa --aligner=bowtie

    # Build genome index for RRBS with default parameters specifying the path for bowtie2
    python bs_seeker2-build.py -f genome.fa --aligner=bowtie2 -p ~/install/bowtie2-2.0.0-beta7/ -r

    # Build genome index for RRBS library using bowite2, with fragment lengths ranging [40bp, 400bp]
    python bs_seeker2-build.py -f genome.fa -r -l 40 -u 400 --aligner=bowtie2

    # Build genome index for RRBS library for double-enzyme :
    # MspI (C-CGG) & ApeKI (G-CWGC, where W=A|T, see [IUPAC code](http://www.bioinformatics.org/sms/iupac.html))
    python bs_seeker2-build.py -f genome.fa -r -c C-CGG,G-CWGC --aligner=bowtie

```


* Tips:

> Index built for BS-Seeker2 is different from the index for BS-Seeker 1.
For RRBS, you need to specify "-r" in the parameters. Also, you need to specify LOW_BOUND and UP_BOUND for the range of fragment lengths according your protocol.

> The fragment length is different from read length. Fragments refers to the DNA fragments which you get by size-selection step (i.e. gel-cut oor AMPure beads). Lengths of fragments are supposed to be in a range, such as [50bp,250bp].

> The indexes for RRBS and WGBS are different. Also, indexes for RRBS are specific for fragment length parameters (LOW_BOUND and UP_BOUND).




## 4.3 bs_seeker2-align.py

Module to map reads on 3-letter converted genome.

* Usage :

```bash
	$ bs_seeker2-align.py -h
    Usage: bs_seeker2-align.py {-i <single> | -1 <mate1> -2 <mate2>} -g <genome.fa> [options]

    Options:
      -h, --help            show this help message and exit

      For single end reads:
        -i INFILE, --input=INFILE
                            Input read file (FORMAT: sequences, qseq, fasta,
                            fastq). Ex: read.fa or read.fa.gz

      For pair end reads:
        -1 FILE, --input_1=FILE
                            Input read file, mate 1 (FORMAT: sequences, qseq,
                            fasta, fastq)
        -2 FILE, --input_2=FILE
                            Input read file, mate 2 (FORMAT: sequences, qseq,
                            fasta, fastq)
        -I MIN_INSERT_SIZE, --minins=MIN_INSERT_SIZE
                            The minimum insert size for valid paired-end
                            alignments [Default: 0]
        -X MAX_INSERT_SIZE, --maxins=MAX_INSERT_SIZE
                            The maximum insert size for valid paired-end
                            alignments [Default: 500]

      Reduced Representation Bisulfite Sequencing Options:
        -r, --rrbs          Map reads to the Reduced Representation genome
        -c pattern, --cut-site=pattern
                            Cutting sites of restriction enzyme. Ex: MspI(C-CGG),
                            Mael:(C-TAG), double-enzyme MspI&Mael:(C-CGG,C-TAG).
                            [Default: C-CGG]
        -L RRBS_LOW_BOUND, --low=RRBS_LOW_BOUND
                            Lower bound of fragment length (excluding C-CGG ends)
                            [Default: 20]
        -U RRBS_UP_BOUND, --up=RRBS_UP_BOUND
                            Upper bound of fragment length (excluding C-CGG ends)
                            [Default: 500]

      General options:
        -t TAG, --tag=TAG   [Y]es for undirectional lib, [N]o for directional
                            [Default: N]
        -s CUTNUMBER1, --start_base=CUTNUMBER1
                            The first cycle of the read to be mapped [Default: 1]
        -e CUTNUMBER2, --end_base=CUTNUMBER2
                            The last cycle of the read to be mapped [Default: 200]
        -a FILE, --adapter=FILE
                            Input text file of your adaptor sequences (to be
                            trimmed from the 3'end of the reads, ). Input one seq
                            for dir. lib., twon seqs for undir. lib. One line per
                            sequence. Only the first 10bp will be used
        --am=ADAPTER_MISMATCH
                            Number of mismatches allowed in adapter [Default: 0]
        -g GENOME, --genome=GENOME
                            Name of the reference genome (should be the same as
                            "-f" in bs_seeker2-build.py ) [ex. chr21_hg18.fa]
        -m NO_MISMATCHES, --mismatches=NO_MISMATCHES
                            Number(>=1)/Percentage([0, 1)) of mismatches in one
                            read. Ex: 4 (allow 4 mismatches) or 0.04 (allow 4%
                            mismatches) [Default: 4]
        --aligner=ALIGNER   Aligner program for short reads mapping: bowtie,
                            bowtie2, soap, rmap [Default: bowtie]
        -p PATH, --path=PATH
                            Path to the aligner program. Detected:
                            bowtie: /Install/bowtie-1.1.2/
                            bowtie2: /weilongguo/Install/bowtie2-master/
        -d DBPATH, --db=DBPATH
                            Path to the reference genome library (generated in
                            preprocessing genome) [Default: /Install/BSseeker2/bs_utils/reference_genomes]
        -l INT, --split_line=INT
                            Number of lines per split (the read file will be split
                            into small files for mapping. The result will be
                            merged. [Default: 4000000]
        -o OUTFILE, --output=OUTFILE
                            The name of output file [INFILE.bs(se|pe|rrbs)]
        -f FORMAT, --output-format=FORMAT
                            Output format: bam, sam, bs_seeker1 [Default: bam]
        --no-header         Suppress SAM header lines [Default: False]
        --temp_dir=PATH     The path to your temporary directory [Detected:
                            /tmp/]
        --XS=XS_FILTER      Filter definition for tag XS, format X,Y. X=0.8 and
                            y=5 indicate that for one read, if #(mCH sites)/#(all
                            CH sites)>0.8 and #(mCH sites)>5, then tag XS:i:1; or
                            else tag XS:i:0. [Default: 0.5,5]
        --XSteve            Filter definition for tag XS, proposed by Prof. Steve
                            Jacobsen, reads with at least 3 successive mCHH will
                            be labeled as XS:i:1,useful for plant genome, which
                            have high mCHG level. Will override --XS option.
        -M FileName, --multiple-hit=FileName
                            File to store reads with multiple-hits
        -u FileName, --unmapped=FileName
                            File to store unmapped reads
        -v, --version       show version of BS-Seeker2

      Aligner Options:
        You may specify any additional options for the aligner. You just have
        to prefix them with --bt- for bowtie, --bt2- for bowtie2, --soap- for
        soap, --rmap- for rmap, and BS-Seeker2 will pass them on. For example:
        --bt-p 4 will increase the number of threads for bowtie to 4, --bt--
        tryhard will instruct bowtie to try as hard as possible to find valid
        alignments when they exist, and so on.
```


* Examples :

```bash

    # WGBS library ; alignment mode, bowtie ; map to WGBS index
    python bs_seeker2-align.py -i WGBS.fa --aligner=bowtie -o WGBS.bam -f bam -g genome.fa

    # WGBS library ; alignment mode, bowtie2-local ; map to WGBS index
    python bs_seeker2-align.py -i WGBS.fa --aligner=bowtie2 -o WGBS.bam -f bam -g genome.fa

    # WGBS library ; alignment mode, bowtie2-end-to-end ; map to WGBS index
    python bs_seeker2-align.py -i WGBS.fa -m 3 --aligner=bowtie2 -o WGBS.bam -f bam -g genome.fa --bt2--end-to-end

    # RRBS library ; alignment mode, bowtie ; map to RR index
    python bs_seeker2-align.py -i RRBS.fa --aligner=bowtie -o RRBS.bam -g genome.fa -r -a adapter.txt

    # RRBS library ; alignment mode, bowtie ; map to WG index
    python bs_seeker2-align.py -i RRBS.fa --aligner=bowtie -o RRBS.bam -g genome.fa -a adapter.txt

    # RRBS library ; alignment mode, bowtie2-end-to-end ; map to WG index
    python bs_seeker2-align.py -i RRBS.fa --aligner=bowtie -o RRBS.bam -g genome.fa -a adapter.txt --bt2--end-to-end

    # Align from qseq format for RRBS with bowtie, specifying lengths of fragments ranging [40bp, 400bp]
    python bs_seeker2-align.py -i RRBS.qseq --aligner=bowtie -o RRBS.bam -f bam -g genome.fa -r --low=40 --up=400 -a adapter.txt

    # The parameters '--low' and '--up' should be the same with corresponding parameters when building the genome index
    # WGBS library ; alignment mode, bowtie ; map to WGBS index; use 8 threads for alignment
    # BS-Seeker2 will run TWO bowtie instances in parallel.
    python bs_seeker2-align.py -i WGBS.fa --aligner=bowtie -o WGBS.bam -f bam -g genome.fa --bt-p 4


```


* Input file:

- Adapter.txt (example for single-end WGBS / RRBS) :

            AGATCGGAAGAGCACACGTC

- Adapter.txt (example for paired-end WGBS) :

            <adapter for mate 1>
            <adapter for mate 2>


* Output format:

- SAM format

    Sample:

    ```
        10918   0       chr1    133859922       255     100M    *       0       0       TGGTTGTTTTTGTTATAGTTTTTTGTTGTAGAGTTTTTTTTGGAAAGTTGTGTTTATTTTTTTTTTTGTTTGGGTTTTGTTTGAAAGGGGTGGATGAGTT        *       XO:Z:+FW        XS:i:0  NM:i:3  XM:Z:x--yx-zzzy--y--y--zz-zyx-yx-y--------z------------x--------z--zzz----y----y--x-zyx--------y--------z   XG:Z:-C_CGGCCGCCCCTGCTGCAGCCTCCCGCCGCAGAGTTTTCTTTGGAAAGTTGCGTTTATTTCTTCCCTTGTCTGGGCTGCGCCCGAAAGGGGCAGATGAGTC_AC
    ```

    Format descriptions:

        BS-Seeker2 specific tags:
        XO : orientation, from forward/reverted
        XS : 1 when read is recognized as not fully converted by bisulfite treatment, or else 0
        XM : number of sites for mismatch
                X: methylated CG
                x: un-methylated CG
                Y: methylated CHG
                y: un-methylated CHG
                Z: methylated CHH
                z: un-methylated CHH
        XG : genome sequences, with 2bp extended on both ends, from 5' to 3'
        YR : tag only for RRBS, serial id of mapped fragment
        YS : tag only for RRBS, start position of mapped fragment
        YE : tag only for RRBS, end position of mapped fragment

        Note:
            For reads mapped on Watson(minus) strand, the 10th colum in SAM file is not the original reads but the revered sequences.


- BS_Seeker format

    Sample:

    ```
        read10	 1	+FW	chr1+0000169137	TC_CGGGGGTTATATGAGTGTGACGGCTGTAGCGTTAGGTGACGATGTCATCTCCGCGTTCCAAGCGTTATGTGCGCACTGAGGGACACATCCACGTTCCCGG_GG	CGGGGGTTATATGAGTGTGATGGTTGTAGCGTTAGGTGATGATGTTATTTTTGCGTTTTAAGCGTTATGTGCGTATTGAGGGATATATTTACGTTTTTGA	X-------------------x--y-----X---------x-----z--z-yx-X---zz---X--------X-z-y-------z-z--zz-X---zyx--	0	77	169135	169235
        read102	 1	+FW	chr1+0000169137	TC_CGGGGGTTATATGAGTGTGACGGCTGTAGCGTTAGGTGACGATGTCATCTCCGCGTTCCAAGCGTTATGTGCGCACTGAGGGACACATCCACGTTCCCGG_GG	CGGGGGTTATATGAGTGTGATGGTTGTAGCGTTAGGTGATGATGTTATTTTTGCGTTTTAAGCGTTATGTGCGTATTGAGGGATATATTTACGTTTTTGA	X-------------------x--y-----X---------x-----z--z-yx-X---zz---X--------X-z-y-------z-z--zz-X---zyx--	0	77	169135	169235
        read104	 0	+FW	chr1+0000325341	-C_CGGCAAACACCACGCCCCGCGATATGGCAGGATTCATGCCGACTAATGGAAAACACACCAGATGCTGGAAAGAGATAAAGGAGAGCGTTACTGCAATACT_GT	CGGTAAATATTACGTTTCGCGATATGGTAGGATTTATGTCGATTAATGGAAAATATATCAGATGTTGGAAAGAGATAAAGGAGAGCGTTATTGTAATATT	X--z---z-zz-X-zzyX-X-------y------z---yX--z----------z-z-zY-----y--------------------X----y--z----y-	0	154	325339	325509
        read105	 0	+FW	chr1+0000238994	-C_CGGCCACACAGTGAAAGGCTGGGCTGTGAGAGCTTCGGTGGAAACCAGGCCTTCACCACTTCTTCTCCCTTCAAGCCACACACAGCTGTTGCAAGTTCCG_G-	CGGTTACATAGTGAAAGGTTGGGTTGTGAGAGTTTTGGTGGAAATTAGGTTTTTATTATTTTTTTTTTTTTTAAGTTATATATAGTTGTTGTAAGTTTCG	X--zz-Z-y---------y----y--------z--x--------zy---zz--z-zz-z--z--z-zzz--z---zz-z-z-y--y-----z-----yX-	0	118	238992	239093
    ```

    Format descriptions:

        (1) Read ID (from the header columns in seq/fastq/qseq/fasta file, or a serial number of the original input)
        (2) Number of mismatches between the genomic seq and the BS read list in columns 6 and 7. The bisulfite converted sites between read Ts to genomic Cs are not included.
        (3) The strand which the read may be from (+FW, +RC, -RC, -FW)
        (4) The coordinate of the mapped position, indicating [the chromosome], [the mapped strand ("+" or "-")], and [the 0-based, 5'-end coordinate of the mapped genomic sequence on the Watson strand].
        (5) BS read sequences from 5' to 3': if the reads are uniquely mapped as they were FW reads, the original reads are shown. If the reads are uniquely mapped as they were RC reads, their reverse complements are shown.
        (6) Summarized sequence of methylated sites: the methylated CG/CHG/CHH sites are marked as X/Y/Z (upper case), whereas the unmethylated CG/CHG/CHH sites are marked as x/y/z (lower case). This column is summarised directly from Columns 6 and 7.
        (7) XS tag, 1 when read is recognized as not fully converted by bisulfite treatment, or else 0
        (8) my_region_serial, tag only for RRBS, serial id of mapped fragment
        (9) my_region_start, tag only for RRBS, start position of mapped fragment
        (10) my_region_end, tag only for RRBS, end position of mapped fragment


* Tips:

- Removing adapter is recommended.

	If you don't know what's your parameter, please ask the person who generate the library for you.

	If you are too shy to ask for it, you can try to de novo motif finding tools (such as [DME](http://cb1.utdallas.edu/dme/index.htm) and [MEME](http://meme.nbcr.net/meme/cgi-bin/meme.cgi)) find the enriched pattern in 1000 reads.

	Of course, you can also use other tools (such as [cutadapt](http://code.google.com/p/cutadapt/) ) to remove adaptor first.

- It's always better to use a wider range for fragment length.

	For example, if 95% of reads come from fragments with length range [50bp, 250bp], you'd better choose [40bp, 300bp].

- Fewer mismatches for the 'local alignment' mode.

    As the 'local alignment', the bad sequenced bases are usually trimmed, and would not be considered by the parameter "-m".
    It is suggested to user fewer mismatches for the 'local alignment' mode.



## 4.4 bs_seeker2-call_methylation.py

This module calls methylation levels from the mapping result.

* Usage:

```bash

	$ bs_seeker2-call_methylation.py -h
    Usage: bs_seeker2-call_methylation.py [options]

    Options:
      -h, --help            show this help message and exit
      -i INFILE, --input=INFILE
                            BAM output from bs_seeker2-align.py
      -d DBPATH, --db=DBPATH
                            Path to the reference genome library (generated in
                            preprocessing genome) [Default: /Install/BSseeker2/bs_utils/reference_genomes]
      -o OUTFILE, --output-prefix=OUTFILE
                            The output prefix to create ATCGmap and wiggle files.
                            Three files (ATCGmap, CGmap, wig) will be generated if
                            specified. Omit this if only to generate specific
                            format.
      --sorted              Specify when the input bam file is already sorted, the
                            sorting step will be skipped [Default: False]
      --wig=OUTFILE         Filename for wig file. Ex: output.wig, or
                            output.wig.gz. Can be overwritten by "-o".
      --CGmap=OUTFILE       Filename for CGmap file. Ex: output.CGmap, or
                            output.CGmap.gz. Can be overwritten by "-o".
      --ATCGmap=OUTFILE     Filename for ATCGmap file. Ex: output.ATCGmap, or
                            output.ATCGmap.gz. Can be overwritten by "-o".
      -x, --rm-SX           Removed reads with tag 'XS:i:1', which would be
                            considered as not fully converted by bisulfite
                            treatment [Default: False]
      --rm-CCGG             Removed sites located in CCGG, avoiding the bias
                            introduced by artificial DNA methylation status
                            'XS:i:1', which would be considered as not fully
                            converted by bisulfite treatment [Default: False]

RELEASE_NOTES
==========================================
    THE RELEASE NOTE OF BS-SEEKER2
==========================================

v2.0.3 - May 19, 2013
1. Fix bug in utils.py
2. Add UPDATE file

------------------------------------------

v2.0.4 - July 29, 2013
1. Update README.md (Definition of ATCGmap format)
2. Fix the "None" output when no parameter
3. Fix error report when using wrong path for short reads aligner
4. MIT License

------------------------------------------

V2.0.5 - Nov 31, 2013
1. Fix bug in "bs_single_end.py" for "numbers_premapped_lst[0] += len(Unique_FW_C2T)"
2. Make the default aligner to Bowtie
3. Update the README file
4. "-r N" in call-methylation will only control Wiggle file
5. Change the output of pair-end mapping to the standard format of SAM file (the QNAME and FLAG fields)
6. The artificially synthesized bases at the ends of RRBS fragments are removed when removing the adapters
7. Add the total number of bases of uniquely mapped reads to the STDOUT
8. Support "end-to-end" alignment mode in the Galaxy version
9. Support input file in both TEXT or Compressed (.gz) format
10. "-m" allow both INT and FLOAT as input. Ex: '-m 5' or '-m 0.05'
11. Add mode information in output

------------------------------------------

V2.0.6 - Jun 9, 2014
1. Add/Revise option "-M"/"-U" for output multiple-/un-mapped reads
2. Fix a bug in "bs_seeker2-call_methylation.py".
    In the version V2.0.5, CGmap file will show A, T, C, G sites; in this version and following,
    CGmap file will only show the C and G sites (with enough coverage).
3. Fix a bug in the output format "BS_Seeker_1"
    In the v2.0.5, the "BS_Seeker_1" format has 3 more columns than previous format defined in "BS Seeker (1)".
    In this version, the "BS_Seeker_1" format is revised to be consistent with the one in "BS Seeker (1)".
    New definition of "BS_Seeker_1" format is add in "README.md".


------------------------------------------

V2.0.7 - Sep 4, 2014
1. [bs_seeker2-call_methylation.py] : add the parameter "--rm-CCGG", in order to avoid the biased caused by
    artifact methylated C at the ends of fragments
2. [bs_seeker2-align.py] : Fix bugs under related with the parameter "--unmapped".
3. [bs_seeker2-call_methylation.py] : round the "methylation level" to 2-digits in all the output (ATCGmap, CGmap,
    wiggle) formats

------------------------------------------

V2.0.8 - Sep 7, 2014
1. [bs_align/*.py] : Fix bug with "-m" and "-u" when splitting large input files internally.
2. [bs_seeker2-align.py] : Update the help information

------------------------------------------

V2.0.9 - Dec 30, 2014
1. [bs_seeker2-call_methylation] : Add the parameter "--rm-overlap", the overlapped region of two mates in
    paired-end library will be counted only once.
2. [bs_seeker2-call_methylation] : All the first line "type wiggle_0" for wiggle output
3. [bs_seeker2-call_methylation] : Modified codes for speed improvements
4. [bs_seeker2-call_methylation] : Add the parameter"--sorted"
5. [bs_align/*.py] : Fix input empty line problem


------------------------------------------

V2.0.10 - Nov 5, 2015
1. [bs_seeker2-align.py] : Fix bug with "-s" and "-e"
2. [bs_seeker2-call_methylation.py] : Fix error with pysam version higher than 0.8.0

------------------------------------------

V2.1.0 - Dec 9, 2016
1. fixed bug related with pysam version
2. fixed bug related with output information

------------------------------------------

V2.1.1 - Apr. 13, 2017
1. [bs_seeker2-call_methylation.py] : fixed bug related with pysam version
2. [bs_seeker2-call_methylation.py] : modified output strategies

------------------------------------------

V2.1.2 - Jun. 23, 2017
1. [bs_seeker2-align.py] : Add option "--XSteve", alternative filtering method proposed by Prof. Steve Jacobson


------------------------------------------
V2.1.3 - Oct. 25, 2017
1. [bs_seeker2-call_methylation.py] : Add option "-d", allow user-defined highest coverage.


------------------------------------------
V2.1.4 - Nov. 13, 2017
1. [bs_align/bs_align_util.py] : fix bug in function of getting methylation levels for one read

------------------------------------------
V2.1.5 - Dec. 21, 2017
1. [bs_seeker2-build.py] : fix bug in finding genome file

------------------------------------------
V2.1.6 - Feb. 1, 2018
1. [bs_seeker2-build.py] : support built-in option for bowtie-build and bowtie2-align

------------------------------------------
V2.1.7 - Oct. 23, 2018
1. [bs_utils/utils.py] : fix a bug in chromosome name such as converting '.' to '_'

------------------------------------------
V2.1.8 - Oct. 30, 2018
1. [ThreadFasta & ChangeCoordicate] : provide solution for solving the large number of scaffold problem


bs_seeker2-build.py
#!/usr/bin/env python

import os
from optparse import OptionParser, OptionGroup
from bs_index.wg_build import *
from bs_index.rrbs_build import *
from bs_utils.utils import *


if __name__ == '__main__':
    #
    parser = OptionParser()
    #
    parser.add_option("-f", "--file", dest="filename",
                      help="Input your reference genome file (fasta)",
                      metavar="FILE")
    parser.add_option("--aligner", dest="aligner",
                      help="Aligner program to perform the analysis: " +
                           ', '.join(supported_aligners) + " [Default: %default]",
                      metavar="ALIGNER", default = BOWTIE)
    parser.add_option("-p", "--path", dest="aligner_path",
                      help="Path to the aligner program. Detected: " +' '*70+
                           '\t'.join(('%s: %s '+' '*70) % (al, aligner_path[al]) for al in sorted(supported_aligners)),
                      metavar="PATH")
    parser.add_option("-d", "--db", type="string", dest="dbpath",
                      help="Path to the reference genome library (generated in preprocessing genome) [Default: %default]",
                      metavar="DBPATH", default = reference_genome_path)

    parser.add_option("-v", "--version", action="store_true", dest="version",
                      help="show version of BS-Seeker2", default=False)
    #
    # RRBS options
    rrbs_opts = OptionGroup(parser, "Reduced Representation Bisulfite Sequencing Options",
                                "Use this options with conjuction of -r [--rrbs]")
    rrbs_opts.add_option("-r", "--rrbs", action="store_true", dest="rrbs",
                         help = 'Build index specially for Reduced Representation Bisulfite Sequencing experiments. '
                                'Genome other than certain fragments will be masked. [Default: %default]',
                         default = False)
    rrbs_opts.add_option("-l", "--low",type= "int", dest="low_bound",
                         help="lower bound of fragment length (excluding recognition sequence such as C-CGG) "
                              "[Default: %default]",
                         default = 20)
    rrbs_opts.add_option("-u", "--up", type= "int", dest="up_bound",
                         help="upper bound of fragment length (excluding recognition sequence such as C-CGG ends) "
                              "[Default: %default]",
                         default = 500)
    rrbs_opts.add_option("-c", "--cut-site", type= "string", dest="cut_format",
                         help="Cut sites of restriction enzyme. Ex: MspI(C-CGG), Mael:(C-TAG), double-enzyme MspI&Mael:"
                              "(C-CGG,C-TAG). [Default: %default]",
                         default = "C-CGG")
    parser.add_option_group(rrbs_opts)
    #
    # option group 5
    opt_group = OptionGroup(parser, "Build-in Aligner Options",
        "You may specify any additional options for the build-in aligner. You just have to prefix them with " +
        ', '.join('%s for %s' % (aligner_options_prefixes[aligner], aligner) for aligner in supported_aligners)+
        ', and BS-Seeker2 will pass them on. For example: --bt--threads=4 will increase the number of threads for bowtie-build to 4, '
        '--bt2--large-index will instruct bowtie2-build to use large-index mode for building index, and so on. ')
    parser.add_option_group(opt_group)
    #
    # ----------------------------------------------------------------
    # separate aligner options from BS Seeker options
    aligner_options = {}
    bs_seeker_options = []
    i = 1
    while i < len(sys.argv):
        arg = sys.argv[i]
        m = re.match(r'^%s' % '|'.join('(%s)' % aligner_options_prefixes[al] for al in supported_aligners), arg)
        if m:
            a_opt = arg.replace(m.group(0), '-', 1)
            aligner_options[a_opt] = []
            while i + 1 < len(sys.argv) and sys.argv[i + 1][0] != '-':
                aligner_options[a_opt].append(sys.argv[i + 1])
                i += 1
            #
            if len(aligner_options[a_opt]) == 0:  # if it is a key-only option
                aligner_options[a_opt] = True
                #
        else:
            bs_seeker_options.append(arg)
        #
        i += 1
        #
    #
    (options, args) = parser.parse_args(args=bs_seeker_options)
    #
    #
    #(options, args) = parser.parse_args()
    #
    # if no options were given by the user, print help and exit
    if len(sys.argv) == 1:
        parser.print_help()
        exit(0)
    #
    if options.version :
        show_version()
        exit (-1)
    else :
        show_version()
    #
    rrbs = options.rrbs
    #
    if options.filename is not None :
        fasta_file=os.path.expanduser(options.filename)
    else :
        error("Please specify the genome file (Fasta) using \"-f\"")
    #
    if fasta_file is None:
        error('Fasta file for the reference genome must be supported')
    #
    if not os.path.isfile(fasta_file):
        if os.path.isfile(os.path.join(options.dbpath, fasta_file)):
            # Search for options.dbpath to check if the genome file is stored there.
            fasta_file = os.path.join(options.dbpath, fasta_file)
        else:
            error('%s cannot be found' % fasta_file)
        #
    #
    #
    if options.aligner not in supported_aligners:
        error('-a option should be: %s' % ' ,'.join(supported_aligners)+'.')
    #
    #aligner_options = dict(aligner_options_defaults[options.aligner], **aligner_options)
    #
    # ----------------------
    # Aligner option string, new added on 2018-01-30
    aligner_options_string = lambda : ' %s ' % (' '.join(opt_key +
                 (' ' + ' '.join(map(str,opt_val)) # join all values if the value is an array
                  if type(opt_val) is list else
                        ('' if type(opt_val) is bool and opt_val # output an empty string if it is a key-only option
                         else ' ' +str(opt_val)) # output the value if it is a single value
                 )
                for opt_key, opt_val in aligner_options.iteritems() if opt_val not in [None, False]))
    # -----------------------
    #
    builder_exec = os.path.join(options.aligner_path or aligner_path[options.aligner],
                                {BOWTIE   : 'bowtie-build',
                                 BOWTIE2  : 'bowtie2-build',
                                 SOAP     : '2bwt-builder',
                                 RMAP     : '' # do nothing
                                }[options.aligner])
    #
    build_command = builder_exec + aligner_options_string() + \
                                 { BOWTIE   : ' -f %(fname)s.fa %(fname)s',
                                     BOWTIE2  : ' -f %(fname)s.fa %(fname)s',
                                     SOAP     : ' %(fname)s.fa'
                                 }[options.aligner]
    #
    #
    print "Reference genome file: %s" % fasta_file
    print "Reduced Representation Bisulfite Sequencing: %s" % rrbs
    print "Short reads aligner you are using: %s" % options.aligner
    print "Builder path: %s" % builder_exec
    #
    #---------------------------------------------------------------
    #
    if not os.path.isfile( builder_exec ) :
        error("Cannot file program %s for execution." % builder_exec)
    #
    ref_path = options.dbpath
    #
    if os.path.exists(ref_path):
        if not os.path.isdir(ref_path):
            error("%s must be a directory. Please, delete it or change the -d option." % ref_path)
        #
    else:
        os.mkdir(ref_path)
    #
    if rrbs: # RRBS preprocessing
        rrbs_build(fasta_file, build_command, ref_path, options.low_bound, options.up_bound, options.aligner, options.cut_format)
    else: # Whole genome preprocessing
        wg_build(fasta_file, build_command, ref_path, options.aligner)
    #
#

